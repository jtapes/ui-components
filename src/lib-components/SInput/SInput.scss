@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-input {
  position: relative;
  display: flex;
  align-items: center;
  border-bottom: 1px solid $color--border;
  padding-top: $size--16;
  height: $size--48;
  transition-property: background-color, color, border, box-shadow;
  transition-duration: $animate--time;

  &__placeholder {
    position: absolute;
    transition: $animate--time;

    &--focused {
      @include text-xs;

      color: $color--font-muted;
      transform: translateY(-100%);
    }
  }

  &__inner {
    border: none;
    width: 100%;
    color: $color--font;

    &::placeholder {
      display: block;
      color: $color--font-secondary;
    }

    &--prepended {
      padding-left: $size--8;
    }

    &--appended {
      padding-right: $size--8;
    }
  }

  &__prepend {
    display: flex;
    align-items: center;
    margin-left: $size--4;

    > .svg-icon {
      width: $size--40;
      color: $color--font-secondary;
    }
  }

  &__append {
    display: flex;
    align-items: center;
    margin-right: $size--4;

    > .svg-icon {
      width: $size--40;
      color: $color--font-secondary;
    }
  }

  &__clear {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0 $size--16;
    color: $color--font-muted;
    cursor: pointer;

    &:hover {
      color: $color--primary;
    }
  }

  &:focus,
  &:hover {
    border-bottom: 1px solid $color--primary;
  }

  &:focus-within {
    border-color: transparent;
    box-shadow: 0 3px $color--outline;
  }

  &--readonly {
    pointer-events: none;

    .s-input {
      &__placeholder {
        color: $color--font-muted;
      }

      &__inner {
        color: $color--font-secondary;
        background-color: transparent;
      }
    }

    &.s-input--rounded {
      border-color: $color--background;
      background-color: $color--background;

      .s-input__inner {
        color: $color--font-secondary;
        background-color: $color--background;
      }
    }
  }

  &--disabled {
    pointer-events: none;

    .s-input {
      &__placeholder {
        color: $color--font-muted;
      }

      &__inner {
        color: $color--font-muted;
        background-color: transparent;
      }
    }

    &.s-input--rounded {
      border-color: $color--background;
      background-color: $color--background;

      .s-input__inner {
        background-color: $color--background;
      }
    }
  }

  &--rounded {
    border: 1px solid $color--border;
    border-radius: $size--24;
    padding-top: 0;
    background-color: $color--white;

    .s-input__inner {
      @include text-sm;

      border-radius: $size--24;
      padding: $size--12 $size--16;
    }

    &:focus,
    &:hover {
      border: 1px solid $color--primary-accent;
    }

    &:focus-within {
      border-color: transparent;
      box-shadow: 0 0 1px 3px $color--outline;
    }

    &.s-input--sm {
      height: $size--40;

      .s-input__inner {
        padding: $size--8 $size--20;
      }
    }
  }

  &--error {
    border-color: $color--danger;
  }

  &--warning {
    border-color: $color--attention;
  }

  &--success {
    border-color: $color--success;
  }

  &--info {
    border-color: $color--primary-accent;
  }
}

input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  appearance: none;
  margin: 0;
}
