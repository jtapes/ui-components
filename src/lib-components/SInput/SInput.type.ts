export enum EInputState {
  DISABLED = 'disabled',
  READONLY = 'readonly',
  ERROR = 'error',
  WARNING = 'warning',
  SUCCESS = 'success',
  INFO = 'info'
}
