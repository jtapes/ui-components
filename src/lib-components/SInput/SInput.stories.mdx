import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest'
import {
  Meta,
  Story,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs'
import SInput from '@/lib-components/SInput/SInput.vue'
import SButton from '@/lib-components/SButton/SButton.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

import results from '/.jest-test-results.json'

import Vue from 'vue'
import mask from '@/plugins/directives/mask'

import { SvgMdxArgType } from '@/storybook/components/icon/SvgIconArgType'

<Meta
  title="DESIGN SYSTEM/Atom/Input"
  component={SInput}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    state: {
      control: { type: 'select' },
      options: [
        'disabled',
        'error',
        'readonly',
        'default',
        'warning',
        'success',
        'info'
      ]
    },
    type: {
      options: ['text', 'number'],
      control: { type: 'select' },
      defaultValue: 'text',
    },
    align: {
      control: { type: 'select' },
      options: [
      'left',
      'center',
      'right'
      ]
    },
    rounded: { control: 'boolean', defaultValue: true },
    placeholder: { control: 'text' },
    size: {
      options: [null, 'sm'],
      control: { type: 'select' },
    },
    clearable: { control: 'boolean', defaultValue: false },
    append: SvgMdxArgType,
    prepend: SvgMdxArgType,
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-%2F-Market-%2F-Catalog-%E2%80%A2-Product-Page?node-id=9%3A4568',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Поле ввода (Input)

<Description of={SInput} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SInput, SvgIcon },
  template: `
    <div style="display: flex; justify-content: center">
      <s-input v-model="model" :state="state" :type="type" :align="align" :rounded="rounded" :placeholder="placeholder" :size="size" :clearable="clearable">
        <template #prepend v-if="prepend">
          <svg-icon :name="prepend" :class="prepend.includes('-16') ? 'icon-16' : 'icon-32'" />
        </template>
        <template #append v-if="append">
          <svg-icon :name="append" :class="append.includes('-16') ? 'icon-16' : 'icon-32'" />
        </template>
      </s-input>
    </div>
  `,
  data: () => {
    return {
      model: '',
    }
  },
})

export const MaskTemplate = (args, { argTypes }) => {
  Vue.use(mask)
  return {
    props: Object.keys(argTypes),
    components: { SInput, SvgIcon },
    template: `
      <div style="display: flex; justify-content: center">
        <s-input v-model="model" v-mask="'+7 (###) ###-##-##'" :state="state" :type="type" :align="align" :rounded="rounded" :placeholder="placeholder" :size="size" :clearable="clearable" />
      </div>
    `,
    data: () => {
      return {
        model: '',
      }
    },
  }
}

export const BasicTemplate = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SInput, SButton, SvgIcon },
  template: `
    <div style="display: flex; justify-content: center">
      <s-input v-model="inputValue" :state="state" :type="type" :align="align" :rounded="rounded" :placeholder="placeholder" :size="size" :clearable="clearable">
        <template #append>
          <svg-icon name="search-24-32" class="icon-32" />
        </template>
      </s-input>
    </div>
  `,
  data: () => ({
    inputValue: '',
  }),
})

<Canvas withToolbar withSource="none">
  <Story name="SInput">{Template.bind({})}</Story>
  <Story name="Basic search" args={{ placeholder: 'Искать' }}>
    {BasicTemplate.bind({})}
  </Story>
  <Story name="SInput not rounded" args={{ rounded: false }}>
    {Template.bind({})}
  </Story>
  <Story name="SInput mask">{MaskTemplate.bind({})}</Story>
</Canvas>

## Пример

  <Source
    language="html"
    code={`
<s-input v-model="model" type="text" rounded placeholder="Искать" size="sm" clearable>
  <template #prepend>
    <svg-icon name="plus-24-32" class="icon-32" />
  </template>
  <template #append>
    <svg-icon name="arrow-24-32" class="icon-32" />
  </template>
</s-input>
  `}
/>

  <ArgsTable of={SInput} />

## Маска

  <Canvas withSource="none">
    <div>
      Для создания масок используется директива v-mask. Токены для маски:
      <Source
        language="html"
        code={`
    '#': {pattern: /\d/},
    'X': {pattern: /[0-9a-zA-Z]/},
    'S': {pattern: /[a-zA-Z]/},
    'A': {pattern: /[a-zA-Z]/, transform: v => v.toLocaleUpperCase()},
    'a': {pattern: /[a-zA-Z]/, transform: v => v.toLocaleLowerCase()},
    '!': {escape: true}
    `}
    />
    Пример:
    <Source
      language="html"
      code={`
    <s-input v-model="model" v-mask="'+7 (###) ###-##-##'" />
    `}
    />
  </div>
</Canvas>
