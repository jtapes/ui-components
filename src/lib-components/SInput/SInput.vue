<template>
  <div
    class="s-input"
    :class="[
      `s-input--${type}`,
      state ? `s-input--${state}` : '',
      {
        's-input--rounded': rounded,
        's-input--sm': size === 'sm',
      },
    ]"
  >
    <div v-if="$slots.prepend" class="s-input__prepend">
      <!-- @slot Слот для prepend элементов -->
      <slot name="prepend" />
    </div>
    <input
      ref="sInput"
      :disabled="state === 'disabled'"
      :readonly="state === 'readonly'"
      :value="value"
      :type="typeInput"
      :min="min"
      :max="max"
      class="s-input__inner"
      :class="{
        'ta-c': align === 'center',
        'ta-r' : align === 'right',
        's-input__inner--appended': $slots.append,
        's-input__inner--prepended': $slots.prepend,
      }"
      :placeholder="rounded ? placeholder : ''"
      @input="handleValueChange"
      @focus="inputFocus = true"
      @focusout="inputFocus = false"
      v-on="listeners"
    />
    <span
      v-if="placeholder && !rounded"
      class="s-input__placeholder"
      :class="{ 's-input__placeholder--focused': placeholderFocus }"
      @mousedown.prevent="inputFocus = true"
    >
      {{ placeholder }}
    </span>
    <div
      v-if="clearable && inputFocus && value"
      class="s-input__clear"
      @click="handleInputClear"
    >
      <svg-icon name="times-24-32" class="icon-32" />
    </div>
    <div v-if="$slots.append" class="s-input__append">
      <!-- @slot Слот для append элементов -->
      <slot name="append" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Model, Vue, Watch } from 'vue-property-decorator'
import { EInputState } from '@/lib-components/SInput/SInput.type'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

/** Компонент для ввода данных */
@Component({
  components: {
    SvgIcon,
  }
})
export default class SInput extends Vue {
  inputFocus = false

  /** Тип инпута */
  @Prop({ type: String, default: 'text' })
  type!: string

  /** Максимальная длина */
  @Prop({ type: Number })
  max!: number

  /** Минимальная длина */
  @Prop({ type: Number })
  min!: number

  /** Свойство округления инпута */
  @Prop({ type: Boolean, default: true })
  rounded!: boolean

  /** Размер инпута */
  @Prop(String)
  size?: string

  /** Возможность очитски инпута */
  @Prop({ type: Boolean, default: false })
  clearable!: boolean

  /** Плэйсхолдер */
  @Prop(String)
  placeholder?: string

  /** Выравнивание текста в иннер инпуте */
  @Prop({ type: String, default: 'left' })
  align!: string

  /** V-model для инпута */
  @Model('change')
  value!: string | number | null

  /** Состояние ("disabled", "error", "readonly") */
  @Prop({
    type: String,
    validator(value: any): boolean {
      return !value || Object.values(EInputState).includes(value)
    },
  })
  state?: EInputState

  $refs!: {
    sInput: HTMLInputElement
  }

  @Watch('inputFocus')
  handleInputFocus(isFocused: boolean): void {
    if (isFocused) {
      this.$refs.sInput.focus()
    }
  }

  get listeners() {
    return {
      ...this.$listeners,
      change: this.handleValueChange,
    }
  }

  get typeInput(): string | null {
    return this.type === 'number' ? null : this.type
  }

  get placeholderFocus(): string | number | boolean | null {
    return this.value || this.inputFocus
  }

  handleValueChange(event: InputEvent): void {
    const target = event.target as HTMLInputElement
    if (target) {
      const value = target.value
      this.$emit('change', value)
    }
  }

  handleInputClear(): void {
    this.$emit('change')
  }
}
</script>
