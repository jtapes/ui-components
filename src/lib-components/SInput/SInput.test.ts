import { mount } from '@vue/test-utils'
// import { fireEvent } from '@testing-library/dom'
import SInput from '@/lib-components/SInput/SInput.vue'

describe('SInput', () => {
  test('Snapshot', () => {
    const wrapper = mount(SInput)
    expect(wrapper.element).toMatchSnapshot()
  })
  test('$emit() работает', async () => {
    const wrapper = mount(SInput)

    const input = wrapper.find('input')
    await input.setValue('test text')

    expect(wrapper.emitted('change')).toBeTruthy()
  })
  test('$emit() работает', async () => {
    const wrapper = mount(SInput, {
      attachTo: document.body,
      propsData: {
        clearable: true,
        value: 'text',
      },
    })

    const input = wrapper.find('input')
    await input.setValue('test text')
    await input.trigger('focus')

    const clearButton = wrapper.find('.s-input__clear')
    await clearButton.trigger('click')

    const emitted = wrapper.emitted('change')

    if (emitted) {
      expect(emitted.length).toBeGreaterThan(1)
      expect(emitted[1]).toEqual([])
    }
  })
  // test('Test not rounded input', () => {
  //   const wrapper = mount(SInput, {
  //     propsData: {
  //       rounded: false,
  //       placeholder: 'Введите',
  //     },
  //   })
  //
  //   const input = wrapper.find('input')
  //   fireEvent.focus(input.element)
  //   expect(wrapper.vm.$data.inputFocus).toBeTruthy()
  // })
})
