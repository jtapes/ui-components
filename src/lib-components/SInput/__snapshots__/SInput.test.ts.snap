// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SInput Snapshot 1`] = `
<div
  class="s-input s-input--text s-input--rounded"
>
  <!---->
   
  <input
    class="s-input__inner"
    type="text"
  />
   
  <!---->
   
  <!---->
   
  <!---->
</div>
`;
